File: Leetcode/FindAllPeopleWithSecret.rs
// https://leetcode.com/problems/find-all-people-with-secret

use std::cmp::Reverse;
use std::collections::BinaryHeap;

struct Solution {}

impl Solution {
    pub fn find_all_people(n: i32, meetings: Vec<Vec<i32>>, first_person: i32) -> Vec<i32> {
        fn dijkstra(graph: &Vec<Vec<(usize, i32)>>, start: usize) -> Vec<Option<(usize, i32)>> {
            let mut ans = vec![None; graph.len()];
            let mut prio = BinaryHeap::new();

            for (new, time) in &graph[start] {
                match ans[*new] {
                    Some((_, new_time)) if new_time <= *time => {}
                    _ => {
                        ans[*new] = Some((start, *time));
                        prio.push(Reverse((*time, new, start)));
                    }
                }
            }

            while let Some(Reverse((secret_time, new, prev))) = prio.pop() {
                match ans[*new] {
                    Some((p, d)) if p == prev && d == secret_time => {}
                    _ => continue,
                }

                for (next, time) in &graph[*new] {
                    match ans[*next] {
                        Some((_, next_secret_time))
                            if secret_time > *time || next_secret_time <= *time => {}
                        None if secret_time > *time => {}
                        _ => {
                            ans[*next] = Some((*new, *time));
                            prio.push(Reverse((*time, next, *new)));
                        }
                    }
                }
            }

            ans
        }

        let (n, first_person) = (n as usize, first_person as usize);

        let mut adj = vec![vec![]; n + 1];
        adj[0].push((first_person, 0));
        adj[first_person].push((0, 0));
        for meeting in meetings.into_iter() {
            adj[meeting[0] as usize].push((meeting[1] as usize, meeting[2]));
            adj[meeting[1] as usize].push((meeting[0] as usize, meeting[2]));
        }

        dijkstra(&adj, 0)
            .into_iter()
            .enumerate()
            .filter_map(|x| x.1.is_some().then_some(x.0 as i32))
            .collect()
    }
}

File: Leetcode/FindAllPeopleWithSecret.cpp
// https://leetcode.com/problems/find-all-people-with-secret

class Solution {
   public:
    vector<int> findAllPeople(int n,
                              vector<vector<int>>& meetings,
                              int firstPerson) {
        vector<vector<pair<size_t, int>>> connections(
            n, vector<pair<size_t, int>>());

        connections[0].push_back({firstPerson, 0});
        for (const auto& meeting : meetings) {
            connections[meeting[0]].push_back({meeting[1], meeting[2]});
            connections[meeting[1]].push_back({meeting[0], meeting[2]});
        }

        vector<int> know_at;
        Solution::dijkstra(0, connections, know_at);

        vector<int> ans;
        for (size_t i = 0; i < (size_t)n; ++i) {
            if (know_at[i] != std::numeric_limits<int>::max()) {
                ans.push_back(i);
            }
        }

        return ans;
    }

    void dijkstra(size_t s,
                  const vector<vector<pair<size_t, int>>>& connections,
                  vector<int>& know_at) {
        constexpr auto INF = std::numeric_limits<int>::max();

        size_t n = connections.size();
        know_at.assign(n, INF);
        know_at[s] = -1;

        set<pair<int, size_t>> st;
        st.insert({know_at[s], s});

        while (!st.empty()) {
            size_t v = st.begin()->second;
            st.erase(st.begin());

            for (auto [to, time] : connections[v]) {
                if (know_at[v] <= time && time < know_at[to]) {
                    auto it = st.find({know_at[to], to});
                    if (it != st.end()) {
                        st.erase(it);
                    }

                    know_at[to] = time;
                    st.insert({know_at[to], to});
                }
            }
        }
    }
};